### ISSUE TITLE
Handle nested structs (for D1)

### ISSUE THREAD

**Issue by daym** [daym/allwinner-register-interface-extractor#5]
Right now, nested struct are not inferred. Example: `0x0050+N*0x0100+P*0x0004(N=0–1)(P=0–3)`

Symptom:

```
  File "/home/dannym/src/allwinner-register-interface-extractor/phase3.py", line 1187, in infer_register_instance_structure
    register_offset = eval(spec[len("Offset:"):].strip(), eval_env)
  File "<string>", line 1, in <module>
NameError: name 'P' is not defined
```

Workaround for the `P`:

```
diff --git a/phase3.py b/phase3.py
index c968b5c..140be89 100644
--- a/phase3.py
+++ b/phase3.py
@@ -1181,6 +1181,7 @@ for module in root_dnode.children:
                       eval_env["N"] = N
                       eval_env["n"] = N
                       eval_env["x"] = N # FIXME remove?
+                      eval_env["P"] = N # FIXME remove?
                       register_offset = eval(spec[len("Offset:"):].strip(), eval_env)
                       common_vars_registers[key][register.name].append(register_offset)
               else:
```

This workaround is REALLY bad--it basically substitutes a random value for `P`. The resulting SVD needs manual editing around that section.

What this is for is there are formulas (for the offsets) in the PDFs and some formulas are of the form `a N + b P + c` (where `a` and `b` and `c` are constants) where it's not possible to know whether this is meant as an array of `b`-sized items of an array of `a`-sized items, or an array of `a`-sized items of an array of `b`-sized items. That's why this needs manual intervention right now.

Furthermore, sometimes it has some outliers, See for example D1.pdf `MSGBOX_MSG_STATUS_REG` and around it on page 344 (visible page number) vs `MSGBOX_DEBUG_REG` also on that page.

Patches welcome--but I think it's hard to infer the actual structure from this.

If we wanted to do that, there would be some kind of formula parser, then it would have to normalize the result into a standard polynomial and then we would have to determine whether `a` or `b` are bigger. Then prefer the bigger value for the outer structure and the smaller value for the inner structure. Then add some ability to actually have nested `cluster`s--which it right now doesn't.

All that for four registers :(

Or just somehow pattern match it and hardcode the replacement result.

**Comment by daym** [comment]
Implemented in master, commit c1f2ef8f073f22d8003ac44bd19b7057ffa0dd3b.

**Comment by kassane** [comment]
> Implemented in master, commit [c1f2ef8](https://github.com/daym/allwinner-register-interface-extractor/commit/c1f2ef8f073f22d8003ac44bd19b7057ffa0dd3b).

- https://github.com/daym/allwinner-register-interface-extractor/issues/4#issuecomment-1224073486
